### PR TITLE
feat: Substrait logical plan

### LINKED ISSUE #532: Support expressing DF logical expr in substrait
### What type of enhancement is this?

API improvement

### What does the enhancement do?

The current substrait <-> DF logical plan doesn't support logical expr, thus filters in `TableScan` cannot be passed to downstream consumer. But on an optimized logical plan, `Filter` plan might be pushed down to `TableScan`, which makes it cannot be simply ignored otherwise the result is wrong.

### Implementation challenges

`DataType` and `Schema` are supported. For an MVP stage, I suppose support selection expr is enough. Let's see if others are easy to add too. This issue only tracks the functionality needed by "executing `TableScan` correctly"

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Use Substrait logical plan to execute query between Frontend and Datanode, in distributed mode. Formerly we are converting table scan to SQL, which is not elegant.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

#682 

#597 

#532

### CODE REVIEW

File: src/datanode/src/instance/grpc.rs [hunk 1]
```diff
@@ -150,10 +150,17 @@ impl Instance {
...
-        let logical_plan = logical_plan_converter
-            .decode(plan_bytes.as_slice())
+    async fn execute_logical(&self, plan: LogicalPlanProto) -> Result<Output> {
+        let table = self
+            .catalog_manager
+            .table(&plan.catalog, &plan.schema, &plan.table)
+            .context(CatalogSnafu)?
+            .context(TableNotFoundSnafu {
+                table_name: format!("{}.{}.{}", &plan.catalog, &plan.schema, &plan.table),
+            })?;
```
Comment (waynexia) [inline src/datanode/src/instance/grpc.rs:160]: `DFLogicalSubstraitConvertor` keeps a `CatalogManager` and will do this when translating `TableScan` plan. So I think these lines are redundant.

**Comment by MichaelScofield** [comment]
@waynexia resolved, PTAL

File: src/table/src/metadata.rs [hunk 2]
```diff
@@ -333,9 +333,9 @@ pub struct TableInfo {
     /// Comment of the table.
     #[builder(default, setter(into))]
     pub desc: Option<String>,
-    #[builder(default, setter(into))]
+    #[builder(default = "\"greptime\".to_string()", setter(into))]
     pub catalog_name: String,
-    #[builder(default, setter(into))]
+    #[builder(default = "\"public\".to_string()", setter(into))]
```
Comment (waynexia) [inline src/table/src/metadata.rs:338]: We have defined these literal `DEFAULT_CATALOG_NAME` and `DEFAULT_SCHEMA_NAME`

File: src/common/substrait/src/df_logical.rs [hunk 3]
```diff
@@ -412,6 +476,24 @@ impl DFLogicalSubstraitConvertor {
     }
 }
 
+fn same_schema_without_metadata(lhs: &ArrowSchemaRef, rhs: &ArrowSchemaRef) -> bool {
+    lhs.fields.len() == rhs.fields.len()
+        && lhs.fields.iter().zip(rhs.fields.iter()).all(|(x, y)| {
+            x.name == y.name && x.data_type == y.data_type && x.is_nullable == y.is_nullable
+        })
+}
```
Comment (waynexia) [inline src/common/substrait/src/df_logical.rs:484]: Looks like it has implemented `Eq` trait https://docs.rs/arrow2/0.10.1/arrow2/datatypes/struct.Schema.html#impl-Eq

File: src/common/substrait/src/df_logical.rs [hunk 4]
```diff
@@ -412,6 +476,24 @@ impl DFLogicalSubstraitConvertor {
...
+fn try_convert_df_schema(df_schema: &DFSchemaRef) -> Result<Schema, Error> {
+    #[allow(clippy::needless_borrow)]
+    let arrow_schema: ArrowSchema = (&**df_schema).into();
+
+    let schema: Schema = arrow_schema
+        .try_into()
+        .map_err(BoxedError::new)
+        .context(InternalSnafu)?;
+    Ok(schema)
+}
```
Comment (waynexia) [inline src/common/substrait/src/df_logical.rs:495]: I prefer to implement this as `impl TryFrom<DfSchema> for Schema` like
```rust
impl TryFrom<Arc<ArrowSchema>> for Schema {
    type Error = Error;
```
which is more generic.

And by the way, can `df_schema.as_ref()` replace that `(&**df_schema)`? It's a trait method from `Arc` https://doc.rust-lang.org/std/sync/struct.Arc.html#impl-AsRef%3CT%3E-for-Arc%3CT%3E

**Comment by MichaelScofield** [comment]
@waynexia resolved, PTAL

**Review by waynexia — APPROVED** [review]

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**